tests/AnalyzeSpec.hs

```diff
@@ -2260,3 +2318,132 @@ spec = describe "analyze" $ do
         @model (property (= result FOO))
         FOO)
       |]
+
+  describe "module-scoped properties verify" $ do
+    let okay = [text|
+          (defun okay:string (from:string to:string)
+            (enforce (!= from to) "sender and receive must not be the same")
+            (with-read accounts from { "balance" := from-bal }
+              (with-read accounts to { "balance" := to-bal }
+                (update accounts from { "balance": (- from-bal 1) })
+                (update accounts to   { "balance": (+ to-bal 1) }))))
+          |]
+        bad = [text|
+          (defun bad:string ()
+            (with-read accounts "joel" { "balance" := bal }
+              (update accounts "joel" { "balance": (+ bal 1000000000) })))
+          |]
+        conservesBalance = "(property conserves-balance)"
+
+    expectVerified'  conservesBalance okay
+    expectFalsified' conservesBalance bad
+    expectFalsified' conservesBalance $ T.unlines [ bad, okay ]
+    expectFalsified' conservesBalance $ T.unlines [ okay, bad ]
+
+    expectVerified' "(property conserves-balance {'except: [bad]})" $
+      T.unlines [ okay, bad ]
+    expectVerified' "(property conserves-balance {'only:   [good]})" $
+      T.unlines [ okay, bad ]
+
+    expectVerified'  "(property conserves-balance {'except: []})" okay
+    expectFalsified' "(property conserves-balance {'except: []})" bad
+
+    expectVerified'  "(property conserves-balance {'only:   []    })" bad
+    expectFalsified' "(property conserves-balance {'only:   [bad]})" bad
+
+  describe "read (property)" $ do
+    let code1 = [text|
+          (defun test:object{account} (acct:string)
+            @model (property (= result (read accounts acct 'before)))
+            (read accounts acct))
+          |]
+    expectVerified code1
+
+    -- reading from a different account
+    let code2 = [text|
+          (defun test:object{account} (acct:string)
+            @model (property (= result (read accounts acct 'before)))
```

Comment:
I wonder whether we should definitely use `read`? Just considering how the differing signatures+semantics of `read` between properties and expressions might be confusing to users...

Reply:
`cell-value`?